README.md
# ResistanceGA

### An R package to optimize resistance surfaces using Genetic Algorithms. Both continuous and categorical surfaces can be optimized using these functions. Additionally, it is possible to simultaneously optimize multiple resistance surfaces at the same time to generate novel resistance surfaces. Resistance distances can be calculated as cost distances (least cost path) between points, or as circuit-based resistance distances calculated using CIRCUITSCAPE

To install this package, execute the following commands in R:

    # Install 'devtools' package, if needed
    if(!("devtools" %in% list.files(.libPaths()))) {
        install.packages("devtools", repo = "http://cran.rstudio.com", dep = TRUE) 
    } 

    devtools::install_github("wpeterman/ResistanceGA", build_vignettes = TRUE) # Download package

    library(ResistanceGA) # Loads package and the other dependnecies

If installation with `build_vignettes = TRUE` fails, try installing the `tinytex` R package and then attempt to install `ResistanceGA` again.

Once the package is installed, you can view the 'Getting Started' vignette in R.

------------------------------------------------------------------------

### Other notes

Optimization with CIRCUITSCAPE (v4) is still possible, although not actively supported in the most recent version. If you wish to optimize using CIRCUITSCAPE, it is highly recommended that you install Julia and the CIRCUITSCAPE Julia package. General instructions [**here**](https://petermanresearch.weebly.com/uploads/2/5/9/2/25926970/julia_guide.pdf "Julia Guide"). There is also a `Julia_Guide` vignette with the package now.

------------------------------------------------------------------------

This approach has been developed from the methods first utilized in Peterman et al. (2014). The first formal analysis using ResistanceGA was Ruiz-López et al. (2016). The primary citation for the package is Peterman (2018), Methods in Ecology.

Peterman, W.E., G.M. Connette, R.D. Semlitsch, and L.S. Eggert. 2014. Ecological resistance surfaces predict fine-scale genetic differentiation in a terrestrial woodland salamander. Molecular Ecology 23:2402–2413. [**PDF**](http://petermanresearch.weebly.com/uploads/2/5/9/2/25926970/peterman_et_al._2014--mec.pdf "Peterman et al.")

Peterman, W. E. 2018. ResistanceGA: An R package for the optimization of resistance surfaces using genetic algorithms. Methods in Ecology and Evolution 9, 1638–1647.
<doi:10.1111/2041-210X.12984>. [**PDF**](https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.12984 "MEE Publication")

Ruiz-López, M.J., Barelli, C., Rovero, F., Hodges, K., Roos, C., Peterman, W.E., Ting, N., 2016. A novel landscape genetic approach demonstrates the effects of human disturbance on the Udzungwa red colobus monkey (*Procolobus gordonorum*). Heredity, Ruiz-López 116, 167–176. <https://doi.org/10.1038/hdy.2015.82>


NEWS
ResistanceGA 4.2-10
--------
* Fixed bug that always rescaled categorical variables to the maximum specified resistance value.

ResistanceGA 4.2-9
--------
* Updated imports to accommodate new package requirements.

ResistanceGA 4.2-8
--------
* Fixed bug causing Errors when optimizing multiple categorical surfaces.

ResistanceGA 4.2-7
--------
* Fixed bug in `gdist.prep` that caused an error when not specifying a response variable.

ResistanceGA 4.2-6
--------
* Updated GA.prep code to allow specification of how many unique values can be considered in a categorical surface

ResistanceGA 4.2-5
--------
* Fixed bug that prevented `jl.prep` from being run when `response = NULL`

ResistanceGA 4.2-4
--------
* Updated rescaling of categorical values to ensure maximum resistance <= specified maximum
* Changed default maximum resistance of categorical and continuous surfaces to be 1,000

ResistanceGA 4.2-3
--------
* Updated rescaling of categorical values to ensure maximum resistance <= specified maximum

ResistanceGA 4.2-2
--------
* Updated files so vignettes can be built across platforms [16 December 2021]

ResistanceGA 4.2-1
--------
* Added checks and warning messages to `GA.prep`, `gdist.prep`, and `jl.prep` [3 August 2021]
* Updated README file [3 August 2020]

ResistanceGA 4.2
--------
* Added a new 'Getting Started' vignette.
* Updated code to create paths across platforms [2021 July 29]

ResistanceGA 4.1-0.47
--------
* Updated `all_comb` function to correctly accomodate a specified number of layers without single surface optimization [2021 July 7]
* Updated `GA.prep` function `percent.elite` to conform to new standards in the GA package [2021 July 27]

ResistanceGA 4.1-0.46
--------
* A lot of clean up of code and updating NAMESPACE [2021 April 27]

ResistanceGA 4.1-0.45
--------
* Fixed bug with using experimental single precision option when optimizing with CS Julia [2021 Jan 14]

ResistanceGA 4.1-0.44
--------
* Fixed potential bug in Bootstrap function

ResistanceGA 4.1-0.43
--------
* Corrected directory setting
* Added Julia Vignette

ResistanceGA 4.1-0.42
--------
* Updated internal raster assessment functions

ResistanceGA 4.1-0.41
--------
* Updated default plotting option

ResistanceGA 4.1-0.40
--------
* Ability to turn off diagnostic plotting in SS & MS optim functions (Default = TRUE)

ResistanceGA 4.1-0.39
--------
* Fixed JuliaCall load error

ResistanceGA 4.1-0.38
--------
* 30 April commit

ResistanceGA 4.1-0.37
--------
* Minor code formatting and updates

ResistanceGA 4.1-0.36
--------
* Updated function documentation

ResistanceGA 4.1-0.35
--------
* Updated gaisl for use with all.comb function

ResistanceGA 4.1-0.34
--------
* Updated pairs_to_include handling
* Fixed bug with Run_gdistance

ResistanceGA 4.1-0.33
--------
* Update to AICc calculation
* Fixed bug in temporary file handling when using Julia

ResistanceGA 4.1-0.32
--------
* Bug fixes and and adjustments to scratch directory use

ResistanceGA 4.1-0.31
--------
* Updated scratch directory use

ResistanceGA 4.1-0.30
--------
* Minor bug fixes

ResistanceGA 4.1-0.29
--------
* Updates to accomodate running on linux cluster

ResistanceGA 4.1-0.28
--------
* Added new sample data sets for running examples

ResistanceGA 4.1-0.27
--------
* Updated Run_julia function to export cumulative .asc maps

ResistanceGA 4.1-0.26
--------
* Minor bug fixes to Julia covariate function

ResistanceGA 4.1-0.25
--------
* Covariates can now be used when optimizing with Julia

ResistanceGA 4.1-0.24
--------
* Updated bootstrap function to accomodate select pairs

ResistanceGA 4.1-0.23
--------
* Now possible to optimize using select pairs of individuals or populations when using Julia

ResistanceGA 4.1-0.22
--------
* Updated multisurface optimization to accommodate covariates
* Added catch to optimization functions constraining resistance distance to have a positive relationship with genetic distance

ResistanceGA 4.1-0.21
--------
* Added function to calculate SHNe metrics (see Prunier et al. 2017, Method in Ecology and Evolution)
* It is now possible to conduct single surface optimization with covariates in the MLPE model. This was specifically designed to incorporate SHNe metrics during optimization

ResistanceGA 4.1-0.20
--------
* Updated all combinations function to pass the best single surface results to the multisurface optimization as starting values.

ResistanceGA 4.1-0.19
--------
* Added ability to fit MLPE with the 'mlpe_rga' function using only select pairwise observations

ResistanceGA 4.1-0.18
--------
* Debugging Julia interaction

ResistanceGA 4.1-0.17
--------
* Updated XRJulia handling of results

ResistanceGA 4.1-0.16
--------
* Updated how Julia interacts with system

ResistanceGA 4.1-0.15
--------
* Added ability to use 'XRJulia' as an alternative to 'JuliaCall' when calling Circuitscape Julia.

ResistanceGA 4.1-0.14
--------
* Have paritally integrated method to optimize on select pairs of samples (not yet functional)
* Added a 'monitor' arguement to GA.prep

ResistanceGA 4.1-0.13
--------
* Updated Bootstrap code

ResistanceGA 4.1-0.12
--------
* Updated code allow ResistanceGA to be run in parallel when using CIRCUITSCAPE

ResistanceGA 4.1-0.11
--------
* Updated and fixed bug in bootstrap function
* Added to switch to run Julia silently
* Updated GA dependency

ResistanceGA 4.1-0.10
--------
* Multisurface k-value bug resolved 

ResistanceGA 4.1-0.3 to 4.1-0.9
--------
* Numerous bug fixes and minor updates. Fixed error causing miscalculation of parameters when using multisurface approach.

ResistanceGA 4.1-0.2
--------
* Extensive updates. Have now fully implemented the Islands optimization approach from the GA package (gaisl), as well as the hybrid local optimization features. Setting `gaisl = TRUE` has great potential to vastly speed up optimization.

ResistanceGA 4.1-0.1
--------
* Continued debugging and upgrading of functions

ResistanceGA 4.1-0
--------

Example.R
require(RandomFields)
require(ResistanceGA)
require(ggplot2)


rm(list = ls())
set.seed(12345)
###################################
# Where are ASCII files and batch files going to be written to?
if("ResistanceGA_Example"%in%dir("C:/")==FALSE) dir.create(file.path("C:/", "ResistanceGA_Example")) 

write.dir <- "C:/ResistanceGA_Example/" # Create a directory to write .asc files and results

# Random point and raster parameters
r.dim <- 60 # number of cells on a side
cell.size <- 0.025
min.point <- 0.25*(r.dim*cell.size) # minimum coordinate for generating random points (multiplied by 0.25 to prevent edge effects)
max.point <- (r.dim*cell.size)-min.point # maximum coordinate for generating random points
n <- 49 # Number of "Sample locations" to generate. This example will generate points on a square grid, so choose a number that has an even square root
x <- seq(from=min.point,max.point,by=cell.size) # set x & y range to draw random samples from
y <- seq(from=min.point,max.point,by=cell.size)
# xy <-c(min.point,max.point,min.point,max.point)
# SAMPLE<-rSSI(.05,n=n,win=xy,giveup=1000) 
xy <-cbind(x,y)
SAMPLE<-SpatialPoints(xy)
COORD <- spsample(Spatial(bbox=bbox(SAMPLE)), n, type="regular") # Generate regularly spaced points on a grid
min(lower(pointDistance(COORD,lonlat=F)))
coord.id <-cbind((1:n),COORD@coords)

write.table(coord.id,file=paste0(write.dir,"samples.txt"),sep="\t",col.names=F,row.names=F)

#####################################################################
#####################################################################
# Using random fields create two resistance surfaces
model<-RMexp() +
  RMtrend(mean=10)

grid.vars <- GridTopology(cellcentre.offset=c(cell.size/2, cell.size/2),
                          cellsize=c(cell.size, cell.size),
                          cells.dim=rep(r.dim,2))

rf.sim <- RFsimulate(model, x=grid.vars, n=2) # Simulate 2 surfaces

cont.rf <- raster(rf.sim[1]) # Define the first as a continuous surface
names(cont.rf)<-"cont"
plot(cont.rf)
plot(COORD, pch=16, col="blue", add=TRUE) # Add randomly generated points

cat.rf <- raster(rf.sim[2]) # Make the second surface a 3-class categorical surface
names(cat.rf)<-"cat"
cat.cut <- summary(cat.rf) # Define quartiles, use these to define categories

cat.rf[cat.rf<=cat.cut[2]] <- 0
cat.rf[cat.rf>0 & cat.rf<=cat.cut[4]] <- 1
cat.rf[!cat.rf%in%c(0,1)] <- 2
plot(cat.rf)
plot(COORD, pch=16, col="blue", add=TRUE)

##############################
# Make categorical feature class (like a road)
feature <- matrix(0,r.dim,r.dim)
feature[25,] <- 1
feature[,30:31] <- 1
feature <- raster(feature)
extent(feature)<-extent(cat.rf)
plot(feature)
names(feature)<-"feature"
plot(feature)
plot(COORD, pch=16, col="blue", add=TRUE)

# Write original rasters to file for use with CIRCUITSCAPE
writeRaster(cat.rf,filename=paste0(write.dir,"cat.asc"),overwrite=TRUE)
writeRaster(cont.rf,filename=paste0(write.dir,"cont.asc"),overwrite=TRUE)
writeRaster(feature,filename=paste0(write.dir,"feature.asc"),overwrite=TRUE)

#########################################
# Run prep functions
GA.inputs<-GA.prep(ASCII.dir=write.dir,
                   pop.mult=5,
                   min.cat=0,
                   max.cat=500,
                   max.cont=500,
                   run=1) # Only single run selected...THIS WILL NOT OPTIMIZE, done for demostration only

CS.inputs<-CS.prep(n.POPS=n,
                   CS_Point.File=paste0(write.dir,"samples.txt"),
                   CS.exe=paste('"C:/Program Files/Circuitscape/4.0/cs_run.exe"')) # Note that RESPONSE is omittted because it has not been made yet

# Run CIRCUITSCAPE to generate pairwise matrix of effective resistance distance
# Only continuous the surface will affect resistance in the first example
CS.Resist<- Run_CS(CS.inputs=CS.inputs,GA.inputs=GA.inputs,r=cont.rf)

# We add some random noise to the response (i.e. CS resistance output)
NOISE <- rnorm(n=length(CS.Resist),mean=0,(0.05*max(CS.Resist))) # Generate random noise matrix
CS.response<-round((CS.Resist+NOISE),digits=4) # Add random noise to resistance output. Use this as the RESPONSE for single surface testing
plot(CS.response~CS.Resist)
write.table(CS.response,paste0(write.dir,"CS.response.txt"),col.names=F,row.names=F)

# Remake the CS.inputs file, including the newly created CS.Response
CS.inputs<-CS.prep(n.POPS=n,
                   RESPONSE=CS.response,
                   CS_Point.File=paste0(write.dir,"samples.txt"),
                   CS.exe=paste('"C:/Program Files/Circuitscape/4.0/cs_run.exe"'))

# Single surface optimization
SS_optim(CS.inputs=CS.inputs,
         GA.inputs=GA.inputs)





#############################
# Visualize transformation of continuous surface. The first term of the PARM function refers to the shape parameter, and the second term refers to the maximum value parameter.

PLOT.trans(PARM=c(2,10),Resistance="C:/ResistanceGA_Example/cont.asc",equation="Reverse Monomolecular",print.dir="C:/ResistanceGA_Example/Results/Plots/")

# Combine raster surfaces, apply transformation to continuous surface and change values of categorical and feature surfaces
PARM=c(0,50,100,3,2,100,0,100)
# PARM=c(0, # First feature of categorical
#        50, # Second feature of categorical
#        100, # Third feature of categorical
#        3, # Transformation equation for continuous surface
#        2, #  Shape parameter
#        100, # Scale parameter
#        0, # First feature of feature surface
#        100) # Second feature of feature surface

PARM=c(0,50,150,3,2,300,0,500)

# Combine resistance surfaces
Resist<-Combine_Surfaces(PARM=PARM,CS.inputs=CS.inputs,GA.inputs=GA.inputs) # Make Combine_Surfaces so that it can take both an R raster object or read a .asc file

# Generate new CS response surface by using Run_CS
CS.Resist<- Run_CS(CS.inputs=CS.inputs,GA.inputs=GA.inputs,r=Resist)

# Generate some random noise nad add it to the resistance surface
NOISE <- rnorm(n=length(CS.Resist),mean=0,(0.1*max(CS.Resist)))
plot(Resist)

CS.response<-round((CS.Resist+NOISE),digits=4)
plot(CS.response~CS.Resist)
write.table(CS.response,file=paste0(write.dir,"ombined_response.csv"),sep=",",row.names=F,col.names=F)

# Run prep functions
GA.inputs<-GA.prep(ASCII.dir=write.dir,
                   pop.mult=5,
                   min.cat=0,
                   max.cat=500,
                   max.cont=750,
                   run=1) # Only single run selected...THIS WILL NOT OPTIMIZE, done for demostration only

CS.inputs<-CS.prep(n.POPS=n,
                      RESPONSE=CS.response,
                      CS_Point.File=paste0(write.dir,"samples.txt"),
                      CS.exe=paste('"C:/Program Files/Circuitscape/4.0/cs_run.exe"'))

Multi.Surface_optim <-MS_optim(CS.inputs=CS.inputs,GA.inputs=GA.inputs)

mantel(CS.Resist~CS.TRUTH)
plot(CS.Resist~CS.TRUTH)


rm(list=".Random.seed", envir=globalenv()) 
